File: url_audit/runner.py
from typing import Callable, Iterable, List, Optional, Sequence, Tuple, Union

from .checks import behavior_checks as bc
from .checks import content_checks as cc
from .checks import domain_checks as dc
from .checks import llm_checks as lc
from .checks import reputation_checks as rc
from .checks import security_checks as sc
from .utils import CheckResult, resolve_audit_target

StepReturn = Union[CheckResult, Sequence[CheckResult], None]
StepFunc = Callable[[str], StepReturn]
Step = Tuple[str, StepFunc]

VALID_STATUSES = {"PASS", "WARN", "FAIL", "INFO"}


CHECK_STEPS: List[Step] = [
    ("Domain Name Legitimacy", dc.check_domain_legitimacy),
    ("Top-Level Domain (TLD)", dc.check_tld),
    ("WHOIS and Domain Age", dc.check_whois_age),
    ("DNS / Email Records", dc.check_dns_email_records),
    ("Registrar Details Transparency", dc.check_registrar_transparency),
    ("Domain Expiry", dc.check_domain_expiry),
    ("Previous Domain Ownership", dc.check_previous_ownership),
    ("Domain Transfer History", dc.check_domain_transfers),
    ("SSL Validity", sc.check_ssl_validity),
    ("HTTPS Presence", sc.check_https_presence),
    ("Certificate Issuer", sc.check_certificate_issuer),
    ("Security Headers", sc.check_security_headers),
    ("IP Reputation", sc.check_ip_reputation),
    ("Server Geolocation", sc.check_server_geolocation),
    ("Hosting Provider", sc.check_hosting_provider),
    ("Page Load Speed", sc.check_page_load_speed),
    ("Mozilla Observatory", sc.check_mozilla_observatory),
    ("Content Quality", cc.check_content_quality),
    ("Spelling Errors", cc.check_spelling_errors),
    ("Brand Consistency", cc.check_brand_consistency),
    ("Contact Information", cc.check_contact_info),
    ("About / Privacy", cc.check_about_privacy),
    ("Too-Good Offers", cc.check_too_good_offers),
    ("Logos & Images", cc.check_logo_images),
    ("Broken Links", cc.check_broken_links),
    ("Security Blacklists", rc.check_security_blacklists),
    ("Google Safe Browsing", rc.check_google_safe_browsing),
    ("Search Visibility", rc.check_search_visibility),
    ("Social Mentions", rc.check_social_mentions),
    ("Wayback Machine", rc.check_wayback),
    ("News & Reviews", rc.check_news_reviews),
    ("Blacklists & Email Filters", rc.check_blacklists_email_filters),
    ("Community Feedback", rc.check_user_community_feedback),
    ("Business Directories", rc.check_business_directories),
    ("Redirect Behaviour", bc.check_redirects),
    ("Popups & Downloads", bc.check_popups_downloads),
    ("Suspicious Requests", bc.check_suspicious_requests),
    ("URL Length", bc.check_url_length),
    ("Homoglyph Detection", bc.check_homoglyph),
    ("Email Links", bc.check_email_links),
    ("Mobile Friendliness", bc.check_mobile_friendly),
    ("Ads & Prompts", bc.check_ads_prompts),
    ("AI Content Analysis", lc.check_llm_content_analysis),
]


def total_steps() -> int:
    return len(CHECK_STEPS)


ProgressCallback = Callable[[int, int, str, List[CheckResult]], None]


def _ensure_list(value: StepReturn) -> List[CheckResult]:
    if value is None:
        return []
    if isinstance(value, CheckResult):
        return [value]
    if isinstance(value, Iterable):
        return [item for item in value if isinstance(item, CheckResult)]
    return []


def _normalize_result_status(result: CheckResult) -> CheckResult:
    status = (result.status or "").upper()
    if status in VALID_STATUSES:
        result.status = status
        return result

    evidence = (result.evidence or "").strip()
    if status == "SKIP":
        result.status = "WARN"
        result.evidence = f"{evidence} | converted_from_skip" if evidence else "converted_from_skip"
        return result

    result.status = "WARN"
    result.evidence = f"{evidence} | normalized_status={status}" if evidence else f"normalized_status={status}"
    return result


def _fallback_result(label: str, index: int, reason: str) -> CheckResult:
    return CheckResult(
        1000 + index,
        label,
        "FAIL",
        evidence=reason,
        data={"error": reason},
    )


def _run_step(url: str, step: Step, index: int) -> List[CheckResult]:
    label, func = step
    try:
        step_results = _ensure_list(func(url))
        if not step_results:
            return [_fallback_result(label, index, "check returned no results")]
        return [_normalize_result_status(result) for result in step_results]
    except Exception as exc:  # noqa: BLE001
        return [_fallback_result(label, index, f"check crashed: {type(exc).__name__}: {exc}")]


def run_all_with_context(
    url: str,
    progress_callback: Optional[ProgressCallback] = None,
) -> Tuple[List[CheckResult], dict]:
    target = resolve_audit_target(url)
    audit_url = target.get("resolved_url") or target.get("normalized_url") or (url or "").strip()

    results: List[CheckResult] = []
    steps = total_steps()

    for index, step in enumerate(CHECK_STEPS, start=1):
        step_results = _run_step(audit_url, step, index)
        if not step_results:
            # Defensive invariant guard.
            step_results = [_fallback_result(step[0], index, "invariant breach: no check result")]

        results.extend(step_results)

        if progress_callback:
            label = step_results[0].name if step_results else step[0]
            progress_callback(index, steps, label, step_results)

    return results, target


def run_all(url: str, progress_callback: Optional[ProgressCallback] = None) -> List[CheckResult]:
    results, _ = run_all_with_context(url, progress_callback)
    return results


def summarize(results: List[CheckResult]):
    counts = {"PASS": 0, "WARN": 0, "FAIL": 0, "INFO": 0}
    for result in results:
        status = (result.status or "").upper()
        if status not in counts:
            status = "WARN"
        counts[status] += 1
    return counts

File: frontend/src/lib/audit/runner.ts
import type { CheckResult } from './checks';
import * as C from './checks';

type CheckFn = (url: string) => Promise<CheckResult | CheckResult[]>;

const CHECK_STEPS: Array<{ label: string; fn: CheckFn }> = [
  { label: 'Domain Name Legitimacy', fn: C.checkDomainLegitimacy },
  { label: 'Top-Level Domain (TLD)', fn: C.checkTld },
  { label: 'WHOIS and Domain Age', fn: C.checkWhoisAge },
  { label: 'DNS / Email Records', fn: C.checkDnsEmailRecords },
  { label: 'Registrar Details Transparency', fn: C.checkRegistrarTransparency },
  { label: 'Domain Expiry', fn: C.checkDomainExpiry },
  { label: 'Previous Domain Ownership', fn: C.checkPreviousOwnership },
  { label: 'Domain Transfer History', fn: C.checkDomainTransfers },
  { label: 'SSL Validity', fn: C.checkSslValidity },
  { label: 'HTTPS Presence', fn: C.checkHttpsPresence },
  { label: 'Certificate Issuer', fn: C.checkCertificateIssuer },
  { label: 'Security Headers', fn: C.checkSecurityHeaders },
  { label: 'IP Reputation', fn: C.checkIpReputation },
  { label: 'Server Geolocation', fn: C.checkServerGeolocation },
  { label: 'Hosting Provider', fn: C.checkHostingProvider },
  { label: 'Page Load Speed', fn: C.checkPageLoadSpeed },
  { label: 'Mozilla Observatory', fn: C.checkMozillaObservatory },
  { label: 'Content Quality', fn: C.checkContentQuality },
  { label: 'Spelling Errors', fn: C.checkSpellingErrors },
  { label: 'Brand Consistency', fn: C.checkBrandConsistency },
  { label: 'Contact Information', fn: C.checkContactInfo },
  { label: 'About / Privacy', fn: C.checkAboutPrivacy },
  { label: 'Too-Good Offers', fn: C.checkTooGoodOffers },
  { label: 'Logos & Images', fn: C.checkLogoImages },
  { label: 'Broken Links', fn: C.checkBrokenLinks },
  { label: 'Security Blacklists', fn: C.checkSecurityBlacklists },
  { label: 'Google Safe Browsing', fn: C.checkGoogleSafeBrowsing },
  { label: 'Search Visibility', fn: C.checkSearchVisibility },
  { label: 'Social Mentions', fn: C.checkSocialMentions },
  { label: 'Wayback Machine', fn: C.checkWayback },
  { label: 'News & Reviews', fn: C.checkNewsReviews },
  { label: 'Blacklists & Email Filters', fn: C.checkBlacklistsEmailFilters },
  { label: 'Community Feedback', fn: C.checkUserCommunityFeedback },
  { label: 'Business Directories', fn: C.checkBusinessDirectories },
  { label: 'Redirect Behaviour', fn: C.checkRedirects },
  { label: 'Popups & Downloads', fn: C.checkPopupsDownloads },
  { label: 'Suspicious Requests', fn: C.checkSuspiciousRequests },
  { label: 'URL Length', fn: C.checkUrlLength },
  { label: 'Homoglyph Detection', fn: C.checkHomoglyph },
  { label: 'Email Links', fn: C.checkEmailLinks },
  { label: 'Mobile Friendliness', fn: C.checkMobileFriendly },
  { label: 'Ads & Prompts', fn: C.checkAdsPrompts },
  { label: 'AI Content Analysis', fn: C.checkLlmContentAnalysis },
];

export function totalSteps(): number {
  return CHECK_STEPS.length;
}

const SECTION_RULES: Array<{ section: string; patterns: string[] }> = [
  { section: 'Domain Intelligence', patterns: ['domain', 'whois', 'tld', 'registrar', 'transfer'] },
  { section: 'Security Posture', patterns: ['ssl', 'https', 'certificate', 'ip reputation', 'hosting', 'security', 'observatory'] },
  { section: 'Reputation & Trust', patterns: ['blacklist', 'safe browsing', 'reputation', 'wayback'] },
  { section: 'Behavioural Signals', patterns: ['redirect', 'suspicious', 'url length', 'homoglyph', 'popup', 'ads', 'mobile', 'email link'] },
  { section: 'Content', patterns: ['content', 'spelling', 'brand', 'contact', 'about', 'privacy', 'offer', 'logo', 'broken', 'image'] },
  { section: 'AI Observations', patterns: ['ai', 'content analysis'] },
];

function lookupSection(name: string): string {
  const lower = (name || '').toLowerCase();
  for (const rule of SECTION_RULES) {
    if (rule.patterns.some((p) => lower.includes(p))) return rule.section;
  }
  return 'Additional Checks';
}

function statusToRisk(status: string): string {
  const map: Record<string, string> = { PASS: 'LOW', INFO: 'LOW', WARN: 'MEDIUM', FAIL: 'HIGH', SKIP: 'LOW' };
  return map[(status || '').toUpperCase()] || 'MEDIUM';
}

const STATUS_STYLES: Record<string, { badge: string; icon: string }> = {
  PASS: { badge: 'success', icon: 'verified' },
  WARN: { badge: 'warning', icon: 'report' },
  FAIL: { badge: 'danger', icon: 'dangerous' },
  INFO: { badge: 'info', icon: 'info' },
  SKIP: { badge: 'secondary', icon: 'upcoming' },
};

export interface PreparedResult {
  id: number;
  name: string;
  status: string;
  badge: string;
  icon: string;
  evidence: string;
  details: string;
  data: Record<string, unknown>;
  summary: string;
  risk_level: string;
  section: string;
}

function prepareResult(r: CheckResult): PreparedResult {
  const style = STATUS_STYLES[r.status] || { badge: 'secondary', icon: 'adjust' };
  const section = lookupSection(r.name);
  const risk_level = statusToRisk(r.status);
  let details = '';
  if (r.data && Object.keys(r.data).length) {
    try { details = JSON.stringify(r.data, null, 2); } catch { details = String(r.data); }
  }
  return {
    id: r.id,
    name: r.name,
    status: r.status,
    badge: style.badge,
    icon: style.icon,
    evidence: r.evidence || '',
    details,
    data: r.data || {},
    summary: r.evidence || `Status reported as ${r.status}`,
    risk_level,
    section,
  };
}

function groupBySection(results: PreparedResult[]): Array<{ name: string; checks: PreparedResult[] }> {
  const sections = new Map<string, PreparedResult[]>();
  for (const r of results) {
    if (!sections.has(r.section)) sections.set(r.section, []);
    sections.get(r.section)!.push(r);
  }
  return Array.from(sections.entries()).map(([name, checks]) => ({ name, checks }));
}

export async function runAll(url: string): Promise<{
  results: PreparedResult[];
  grouped_results: Array<{ name: string; checks: PreparedResult[] }>;
  counts: Record<string, number>;
}> {
  const allResults: PreparedResult[] = [];
  const counts: Record<string, number> = { PASS: 0, WARN: 0, FAIL: 0, INFO: 0, SKIP: 0 };

  for (const step of CHECK_STEPS) {
    try {
      const raw = await step.fn(url);
      const list = Array.isArray(raw) ? raw : [raw];
      for (const r of list) {
        const prepared = prepareResult(r);
        allResults.push(prepared);
        counts[prepared.status] = (counts[prepared.status] || 0) + 1;
      }
    } catch (e) {
      const fallback: PreparedResult = {
        id: 999,
        name: step.label,
        status: 'FAIL',
        badge: 'danger',
        icon: 'dangerous',
        evidence: `check crashed: ${String(e).slice(0, 200)}`,
        details: '',
        data: { error: String(e) },
        summary: `check crashed: ${String(e).slice(0, 100)}`,
        risk_level: 'HIGH',
        section: lookupSection(step.label),
      };
      allResults.push(fallback);
      counts['FAIL'] = (counts['FAIL'] || 0) + 1;
    }
  }

  return {
    results: allResults,
    grouped_results: groupBySection(allResults),
    counts,
  };
}
